src/Algebra/Graph/Undirected.hs

```diff
@@ -0,0 +1,870 @@
+fmapR f g = bindR g (vertex . f)
+{-# INLINE fmapR #-}
+
+-- | __Note:__ this does not satisfy the usual ring laws; see 'Graph' for more
+-- details.
+instance Num a => Num (Graph a) where
+    fromInteger = vertex . fromInteger
+    (+)         = overlay
+    (*)         = connect
+    signum      = const empty
+    abs         = id
+    negate      = id
+
+instance Ord a => Eq (Graph a) where
+    (==) = eqR
+
+instance Ord a => Ord (Graph a) where
+    compare = ordR
+
+-- TODO: Find a more efficient equality check.
+-- TODO: Implement toUndirectedAdjacencyMap.
+-- Check if two graphs are equal by converting them to their symmetric
+-- relations.
+eqR :: Ord a => Graph a -> Graph a -> Bool
+eqR x y = toSymmetricRelation x == toSymmetricRelation y
+{-# NOINLINE [1] eqR #-}
+{-# RULES "eqR/Int" eqR = eqIntR #-}
+
+-- Like 'eqR' but specialised for graphs with vertices of type 'Int'.
+-- NOTE: This is not specialised to vertices of type 'Int'. But it's still
+-- here for when 'UndirectedAdjacencyIntMap' is implemented.
+eqIntR :: Graph Int -> Graph Int -> Bool
+eqIntR x y = toSymmetricRelation x == toSymmetricRelation y
+
+-- TODO: Find a more efficient comparison.
+-- TODO: Implement toUndirectedAdjacencyIntMap.
+-- Compare two graphs by converting them to their adjacency maps.
+ordR :: Ord a => Graph a -> Graph a -> Ordering
+ordR x y = compare (toSymmetricRelation x) (toSymmetricRelation y)
+{-# NOINLINE [1] ordR #-}
+{-# RULES "ordR/Int" ordR = ordIntR #-}
+
+-- Like 'ordR' but specialised for graphs with vertices of type 'Int'.
+-- NOTE: This is not specialised to vertices of type 'Int'. But it's still
+-- here for when 'UndirectedAdjacencyIntMap' is implemented.
+ordIntR :: Graph Int -> Graph Int -> Ordering
+ordIntR x y = compare (toSymmetricRelation x) (toSymmetricRelation y)
+
+instance Applicative Graph where
+    pure  = vertex
+    (<*>) = apR
+
+apR :: Graph (a -> b) -> Graph a -> Graph b
+apR f x = bindR f (<$> x)
+{-# INLINE apR #-}
+
+instance Monad Graph where
+    return = pure
+    (>>=)  = bindR
+
+bindR :: Graph a -> (a -> Graph b) -> Graph b
+bindR g f = foldg empty f overlay connect g
+{-# INLINE [0] bindR #-}
+
+instance Alternative Graph where
+    empty = empty
+    (<|>) = overlay
+
+instance MonadPlus Graph where
+    mzero = empty
+    mplus = overlay
+
+-- Help GHC with type inference (direct use of 'coerce' does not 
+-- compile).
+coerce0 :: G.Graph a -> Graph a 
+coerce0 = coerce
+
+-- Help GHC with type inference (direct use of 'coerce' does not
+-- compile).
+coerce1 :: (Coercible a b) => (a -> G.Graph c) -> (b -> Graph c)
+coerce1 = coerce
+
+-- Help GHC with type inference (direct use of 'coerce' does not
+-- compile).
+coerce2 :: (Coercible a b, Coercible c d) => (a -> c -> G.Graph e) -> (b -> d -> Graph e)
+coerce2 = coerce
+
+-- Help GHC with type inference (direct use of 'coerce' does not
+-- compile).
+coerce3 :: (b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> G.Graph a -> b) 
```

Comment:
Let's call this `coerceFold`.

Reply:
Actually, since you use it only once, I think it shouldn't be a separate function. Just add a type signature locally.

Reply:
Okay!